crates/holochain/src/conductor/conductor.rs

```diff
@@ -188,104 +192,123 @@ where
     DS: DnaStore + 'static,
 {
     pub(super) fn cell_by_id(&self, cell_id: &CellId) -> ConductorResult<Arc<Cell>> {
-        let item = self
+        let cell = self
             .cells
-            .get(cell_id)
+            .share_ref(|c| c.get(cell_id).map(|i| i.cell.clone()))
             .ok_or_else(|| ConductorError::CellMissing(cell_id.clone()))?;
-        Ok(item.cell.clone())
-    }
-
-    /// Iterator over all cells, regardless of status.
-    pub(super) fn _all_cells(&self) -> impl Iterator<Item = (&CellId, &Cell)> {
-        self.cells.iter().map(|(id, item)| (id, item.cell.as_ref()))
+        Ok(cell)
     }
 
     /// Iterator over only the cells which are fully running. Generally used
     /// to handle conductor interface requests
-    pub(super) fn running_cells(&self) -> impl Iterator<Item = (&CellId, &Cell)> {
-        self.cells.iter().filter_map(|(id, item)| {
-            if item.is_running() {
-                Some((id, item.cell.as_ref()))
-            } else {
-                None
-            }
+    pub(super) fn running_cell_ids(&self) -> HashSet<CellId> {
+        self.cells.share_ref(|c| {
+            c.iter()
+                .filter_map(|(id, item)| {
+                    if item.is_running() {
+                        Some(id.clone())
+                    } else {
+                        None
+                    }
+                })
+                .collect()
         })
     }
 
     /// Iterator over only the cells which are pending validation. Used to
     /// discover which cells need to be joined to the network.
-    pub(super) fn pending_cells(&self) -> impl Iterator<Item = (&CellId, &Cell)> {
-        self.cells.iter().filter_map(|(id, item)| {
-            if item.is_pending() {
-                Some((id, item.cell.as_ref()))
-            } else {
-                None
-            }
+    pub(super) fn pending_cells<F, R>(&self, f: F) -> Vec<R>
+    where
+        F: Fn((&CellId, &Cell)) -> R,
+    {
+        self.cells.share_ref(|cells| {
+            cells
+                .iter()
+                .filter_map(|(id, item)| {
+                    if item.is_pending() {
+                        Some((id, item.cell.as_ref()))
+                    } else {
+                        None
+                    }
+                })
+                .map(f)
+                .collect()
```

Comment:
If the caller doesn't know that the function is called within a share, they might do some long-running db operation or something within it.

Consider returning an owned `Vec<CellId, Arc<Cell>>` since those are pretty cheap to clone, and the only usage of this is cloning the bits anyways...

Reply:
That makes sense